### PR TITLE
fix(core): disallow delete or create workspaces

### PR DESCRIPTION
**Pull request by outsinre**
### Summary

Currently, we can create new workspaces and also delete/patch/put workspaces (the `default` included). This is not the intended actions.

Define a specific custom route `kong/api/routes/workspaces.lua` to override the workspaces module created by DAO.

**THIS IS ONLY FOR CE**

### Checklist

- [x] The Pull Request has tests
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Issue reference

Fix #_[FTI-5620]_

[FTI-5620]: https://konghq.atlassian.net/browse/FTI-5620?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by ms2008 — COMMENTED** [review]
This is a CE-only issue, so the fix can't be synchronized to the EE. I'm concerned that if it gets accidentally synced to EE someday, it could cause problems (although there are test cases for this in EE that would allow us to catch the bug early).

**Comment by outsinre** [comment]
> This is a CE-only issue, so the fix can't be synchronized to the EE. I'm concerned that if it gets accidentally synced to EE someday, it could cause problems (although there are test cases for this in EE that would allow us to catch the bug early).

By default, it is not synced to EE, unless add the cherry-pick label.

File: kong/api/api_helpers.lua [hunk 1]
```diff
@@ -478,7 +478,9 @@ function _M.attach_new_db_routes(app, routes)
         return method_handler(self, kong.db, handler_helpers)
       end
 
-      methods[method_name] = parse_params(wrapped_handler)
+      if route_path ~= "/workspaces" or method_name ~= "POST" then
```
Comment (tzssangglass) [inline kong/api/api_helpers.lua:481]: what is the purpose of this judgment here?

Comment (outsinre) [reply kong/api/api_helpers.lua:482]: We do not want the lapis to listen for "POST /workspaces" operation.

File: changelog/unreleased/kong/disallow_add_del_workspaces.yml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+message: |
+  Disallow creating or deleting workspaces.
+type: bugfix
+scope: Core
```
Comment (ms2008) [inline changelog/unreleased/kong/disallow_add_del_workspaces.yml:4]:
```suggestion
scope: Admin API
```

**Review by ms2008 — APPROVED** [review]
A reminder to the PR merger: Do **NOT** bring this into EE.

**Review by gszr — CHANGES_REQUESTED** [review]
You can use the `generate_admin_api` flag in the DAO (`true` by default).
```
diff --git a/kong/db/schema/entities/workspaces.lua b/kong/db/schema/entities/workspaces.lua
index 79f45b10d..153eeb57f 100644
--- a/kong/db/schema/entities/workspaces.lua
+++ b/kong/db/schema/entities/workspaces.lua
@@ -7,6 +7,7 @@ return {
   cache_key = { "name" },
   endpoint_key = "name",
   dao          = "kong.db.dao.workspaces",
+  generate_admin_api = false,
 
   fields = {
     { id          = typedefs.uuid },

```

File: changelog/unreleased/kong/disallow_add_del_workspaces.yml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+message: |
```
Comment (gszr) [inline changelog/unreleased/kong/disallow_add_del_workspaces.yml:1]: Also, let's not mention this in the changelog as this has never been a user-visible feature of OSS.

**Comment by outsinre** [comment]
> You can use the `generate_admin_api` flag in the DAO (`true` by default).
> 
> ```
> diff --git a/kong/db/schema/entities/workspaces.lua b/kong/db/schema/entities/workspaces.lua
> index 79f45b10d..153eeb57f 100644
> --- a/kong/db/schema/entities/workspaces.lua
> +++ b/kong/db/schema/entities/workspaces.lua
> @@ -7,6 +7,7 @@ return {
>    cache_key = { "name" },
>    endpoint_key = "name",
>    dao          = "kong.db.dao.workspaces",
> +  generate_admin_api = false,
>  
>    fields = {
>      { id          = typedefs.uuid },
> ```

Brilliant idea.

Would this also disable the `GET /workspaces`? Anyway, since CE does not support workspace, we should also disable the `GET` as well.

Will update.

File: spec/02-integration/04-admin_api/25-add_delete_workspaces_spec.lua [hunk 4]
```diff
@@ -0,0 +1,50 @@
+local helpers = require "spec.helpers"
```
Comment (gszr) [inline spec/02-integration/04-admin_api/25-add_delete_workspaces_spec.lua:1]: Nitpick: maybe name this file accordingly to how others are neme, after the main entity? eg, `spec/02-integration/04-admin_api/25-workspaces_spec.lua`

File: spec/02-integration/04-admin_api/25-add_delete_workspaces_spec.lua [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+    before_each(function()
+      admin_client = helpers.admin_client()
+    end)
+
+    after_each(function()
+      if admin_client then admin_client:close() end
+    end)
+
+    it("disallow creation", function()
```
Comment (gszr) [inline spec/02-integration/04-admin_api/25-add_delete_workspaces_spec.lua:28]:
```suggestion
    it("has no admin api", function()
```

Comment (gszr) [reply spec/02-integration/04-admin_api/25-add_delete_workspaces_spec.lua:28]: IMO a single test case is enough.

**Review by gszr — APPROVED** [review]
Great job!

**Comment by outsinre** [comment]
OSS ONLY!

**Pull request merged**